### PR TITLE
Implementation of the new hash function to BLS12 curves

### PR DESCRIPTION
**Pull request by dishport**
### Project Abstract

Recently, my article https://link.springer.com/article/10.1007/s10623-022-01012-8 was published in the quite prestigious cryptographic journal "Designs, Codes and Cryptography". This article provides a new hash function (indifferentiable from a random oracle) to the subgroup G1 of pairing-friendly elliptic curves BLS12-381 and BLS12-377 (Barreto-Lynn-Scott). These curves and such hash functions are actively used in blockchains, namely in the BLS (Boneh-Lynn-Shacham) aggregate signature. 
According to the web page https://wiki.polkadot.network/docs/learn-keys, the BLS signature will be in particular used in Polkadot.

My hash function is much faster than previous state-of-the-art ones, including the Wahby-Boneh indirect map. For instance, BLS12-377 is defined over a highly 2-adic finite field Fp (of characteristic p), hence the indifferentiable Wahby-Boneh hash function requires to apply twice the slow Tonelli-Shanks algorithm for extracting two square roots in the basic field. In comparison, the new hash function extracts only one cubic root, which can be expressed via one exponentiation in Fp. I have already checked the correctness of my results in the computer algebra system Magma. The project is dedicated to an implementation of the new hash function in Sage.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [ ] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

### CODE REVIEW

**Comment by alxs** [comment]
Thank you for the application @dishport. This looks interesting and I'd probably support an implementation in Rust. However, unfortunately, we can't provide funding for the time you spend learning a programming language. I recommend you apply again when you're ready to get started or, preferably, that you find someone with experience implementing cryptographic functions in Rust to work with you and then apply.

**Comment by dishport** [comment]
> Thank you for the application @dishport. This looks interesting and I'd probably support an implementation in Rust. However, unfortunately, we can't provide funding for the time you spend learning a programming language. I recommend you apply again when you're ready to get started or, preferably, that you find someone with experience implementing cryptographic functions in Rust to work with you and then apply.

Thank you for the rapid response @alxs. I don't ask to pay me anything at the moment. Nevertheless, could you please first consult with your colleagues and/or read my project carefully to have a clear idea if the project is suitable for the grant program? I just want to be sure that my project will definitely be supported in order to be more motivated to learn Rust. And after that, I will apply once again if it is necessary.

**Comment by dishport** [comment]
> In any case, I can't tell you whether your project will be supported or not with certainty as that depends on the committee's decision at the time you apply. 

@alxs, but I created a full application. If I apply again, it will only be necessary to change a few words. Could the committee take a decision for this application ?

> Personally, I think this should be implemented by someone with experience as I mentioned above and the most you should attempt would be a PoC as long as you're just learning a language. Is there any language you're proficient in? What's your experience with programming?

If you know somebody who is ready to implement my hash function, I will help him/her with pleasure to do this. Otherwise, this grant will be a motivation for me to become proficient in the low-level programming. Many years ago, I learned a lot of programming languages in my university, but after that I have only used the computer algebra systems Magma, Sage, because I am an applied researcher rather than a programmer.

**Comment by swasilyev** [comment]
I support that, whose the committee?
In what language? I would even accept good Sage.

**Comment by burdges** [comment]
We've confidence in @dishport on the theory side so frankly I'm thrilled for us to pay him to explore implementation stuff.  If he runs into trouble then maybe we can help him a bit too.  

Are you on matrix @dishport?  After it basically works then it'll be easier for us to get our people to help you with optimizations and benchmarking in matrix, although maybe it'll be even better if we find outside help later.

**Comment by burdges** [comment]
> we can't provide funding for the time you spend learning a programming language

We should make an exception to this rule in this case.  Argument:

It's kinda impossible to get really careful crypto implementation talent, so I'd rather have someone who really knows the theory side trying, than have someone who can code well screw things up in harder to fix ways.  We won't literally be paying for him to learn, so much as paying him for more hours coding under the assumption that he'll just take longer.

**Comment by dishport** [comment]
@swasilyev, @burdges, thanks a lot for your support.

> Are you on matrix @dishport? After it basically works then it'll be easier for us to get our people to help you with optimizations and benchmarking in matrix, although maybe it'll be even better if we find outside help later.

I am not sure what you mean by matrix in the given context. Please clarify.

**Comment by burdges** [comment]
[Element](https://element.io/) is the usual front end for matrix, although it changes names sometimes.  It's basically a more modern XMPP with sane forward secure encryption built in.

**Comment by mmagician** [comment]
@dishport thanks for the application, looks good.
Regarding the concerns raised above: I would simply bundle the entirely of your expenses into one milestone. It seems the research team is in need of such an implementation, and the price would be justified.
Also, please commit already to the programming language. Ideally this would br Rust, but as @swasilyev says we can accept Sage too. I would appreciate it it if you could include that in the milestone itself.
Once you've made the above updates, I'd be happy to go ahead with the application.

**Comment by dishport** [comment]
@mmagician, thanks for your remarks. I updated my project, taking into account all of them. 
 
> Re matrix: it's the matrix protocol, of which Element: https://element.io/ is a popular client.

Why do we need to use the secure protocol Matrix if the implementation is open-source ?

**Comment by alxs** [comment]
Glad to see this kind of response! You seem to have the support of our research team and seeing as you will be building this in a language you're familiar with, none of my concerns applies any longer.

I have an additional request though, as you saw in the application template we usually require a [deliverable table](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#milestone-1-example--implement-substrate-modules) which I think would also help your application. Could you please add it back in and include deliverables 0a-c, and write a brief specification for your implementation? Feel free to refer to functions defined in your article. Potentially you could also split the functionality into different milestones; if it's detailed enough it would also justify a higher price. @burdges are unit tests a thing in Sage and would it make sense to include them?

Using Matrix/Element would make it easier for you to coordinate with our researchers or to ask questions.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application @dishport! I'm happy to approve it, as soon as it contains the table with the deliverables, that Alex mentioned, see https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#milestone-1-example--implement-substrate-modules These are basically the “requirements” of our contracts. Regarding adding the bitcoin address later, this is fine, but it requires you to update the contract with your current github account again and include the address before we can pay you. This is basically our way to prevent fake invoices. I would also be happy to support you to write an additional rust implementation for example as a second milestone for 5k.  And let me know if you have any questions or want to jump on call.

**Comment by dishport** [comment]
@alxs, @Noc2, I added deliverables and a brief specification for the implementation.

> I would also be happy to support you to write an additional rust implementation for example as a second milestone for 5k. 

For now, I would like to limit myself to Sage, because I have experience with it. Perhaps, we will return to this issue after the Sage implementation is ready.

File: applications/Implementation of the new hash function to BLS12 curves.md [hunk 1]
```diff
@@ -113,6 +113,16 @@ https://link.springer.com/article/10.1007/s10623-022-01012-8
...
 - **Costs:** 10,000 USD  
 
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | GNU GPL |
+| 0b. | Documentation | I will provide **inline documentation** of the code. |
+| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. Ie will describe how to run these tests. |
+| 0d. | Docker | I do not intend to deliver this, because Milestone 1 is research oriented. |
+| 0e. | Article | I will cite the implementation in Section 1.1 of my new article https://eprint.iacr.org/2021/1082. I would like to submit this article to the international conference https://fq15.telecom-paris.fr/index.html in Paris if it is not canceled because of COVID-19.
+
```
Comment (alxs) [inline applications/Implementation of the new hash function to BLS12 curves.md:123]:
```suggestion
| 1. | Implementation | Sage implementation of the hash function described below.

```

**Comment by alxs** [comment]
Great! I think it's OK if you amend the application with your payment address when you deliver. Happy to approve it as is.

**Review by alxs — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]

**Comment by Noc2** [comment]
Congratulations and welcome to the Web3 Foundation Grants Program! Please refer to our [Milestone Delivery repository](https://github.com/w3f/Grant-Milestone-Delivery/) for instructions on how to submit milestones and invoices, our [FAQ](https://github.com/w3f/Grants-Program/blob/master/docs/faq.md) for frequently asked questions and the [support section of our README](https://github.com/w3f/Grants-Program/blob/master/README.md#support) for more ways to find answers to your questions. <br/><br/> Before you start, take a moment to read through our [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/announcement-guidelines.md) for all communications related to the grant or make them known to the right person in your organisation. In particular, please <b>don't announce the grant publicly before at least the first milestone of your project has been approved.</b> At that point or shortly before, you can get in touch with us at grantsPR@web3.foundation and we'll be happy to collaborate on an announcement about the work you’re doing.<br/><br/> Lastly, please remember to let us know in case you run into any delays or deviate from the deliverables in your application. You can either leave a comment here or directly request to amend your application via PR. We wish you luck with your project! :rocket:

**Comment by Noc2** [comment]
Fyi: see above our welcome bot failed because of the name of the file: Alex already renamed it, see here: https://github.com/w3f/Grants-Program/blob/master/applications/new_bls12_hash_function.md Just for future reference.

**Comment by Noc2** [comment]
And congratulations!

**Comment by dishport** [comment]
> And congratulations!

@Noc2, thank you! When will my project be reviewed by the committee?

**Comment by alxs** [comment]
@dishport your application was just approved by the [committee](https://github.com/w3f/Grants-Program#w3f-grants-committee) ;)

**Comment by Noc2** [comment]
Just to clarify. Your application is already approved (= merged). Also see https://github.com/w3f/Grants-Program/blob/master/docs/accepted_grant_applications.md#surfing_woman-wave-13---first-quarter-2022

**Comment by Noc2** [comment]
@dishport Thanks for pinging me here. But before we sign an additional grant, the current milestone needs to be accepted first. In order to accept the milestone, you need to fulfil the requirements of the original contract (e.g. unit test) or this contract needs to be amended. If you want to amend the original contract, you can create a PR to this repo to update it. But as part of the PR, you need to justify why the contract needs to be updated and you were unable to fulfil these requirements. The amendment also requires the necessary number of approvals from the committee.

**Comment by dishport** [comment]
@Noc2, my colleague wants to be anonymous. Is it possible to amend the contract without mentioning his name ?

**Comment by Noc2** [comment]
Yes, that’s possible. However, for the payment the invoice team requires a real address and person (via google form).

**Comment by Noc2** [comment]
@dishport just to mention it here. We would still be interested in the rust implementation with tests. Let me know if we can help in any way. In the meantime it would be nice if you could also look at https://github.com/dishport/Indifferentiable-hashing-to-ordinary-elliptic-curves-of-j-0-with-the-cost-of-one-exponentiation/pull/2

**Comment by dishport** [comment]
@Noc2, yesterday my colleague prepared an implementation (with tests) in Rust. I would like to amend the original contract in the near weeks. 

Are you ready to pay even more if he provides a constant-time implementation ? 

And I will also comment on the code as drskalman requests. Last time I have been quite busy to do this.

**Comment by cruikshankss** [comment]
@dishport, if your colleague presents an application that has good business & technological strategies towards an ideal decentralized web3 tech stack, I do not see why we would reject his application. We generally encourage grant applications from Rust implementers who understand cryptographic security, as they are relatively rare! Thank you!

**Comment by dishport** [comment]
> If you want to amend the original contract, you can create a PR to this repo to update it.

Please explain how to do this ? Do I need to create from scratch an absolutely new project as described on the page 
https://github.com/w3f/Grants-Program ? Or is it sufficient to modify Project Abstract on this page ?

**Comment by Noc2** [comment]
@dishport Both would be fine in this case. You can either amend the current contract or apply for a new grant.

**Comment by dishport** [comment]
> Both would be fine in this case. You can either amend the current contract or apply for a new grant.

@Noc2, I commented on my Sage code (https://github.com/dishport/Indifferentiable-hashing-to-ordinary-elliptic-curves-of-j-0-with-the-cost-of-one-exponentiation) in more detail. And I updated a description of the project.

**Comment by cruikshankss** [comment]
Thank you for updating the description! Due to my lack of understanding, the only feedback I can give on the Sage code comment is thank you for adding them.

The description has such a coincidence that you use a "Tonelli-Shanks algorithm" as part of your cryptography because it's so close to my little brother's video game username "tnelli" and mine is "cruikshankss" haha

Personally, good luck with the amendment, etc. We would be honored to have your cryptography in the ecosystem.

**Pull request merged**